internals-js/src/federation.ts

```diff
@@ -476,14 +476,14 @@ export class FederationMetadata {
 
   private fieldUsedPredicate(): (field: FieldDefinition<CompositeType>) => boolean {
     if (!this._fieldUsedPredicate) {
-      const usedFields = Array.from(collectUsedFields(this));
-      this._fieldUsedPredicate = (field: FieldDefinition<CompositeType>) => !!usedFields.find((f) => f.coordinate === field.coordinate);
+      const usedFields = collectUsedFields(this);
+      this._fieldUsedPredicate = (field: FieldDefinition<CompositeType>) => !!usedFields.has(field);
     }
     return this._fieldUsedPredicate;
   }
 
   isFieldUsed(field: FieldDefinition<CompositeType>): boolean {
-    return this.fieldUsedPredicate()(field);
+    return this.fieldUsedPredicate()(field) || this.keysPredicate()(field);
```

Comment:
I still don't understand why that `keysPredicate` check is needed, could you explain?

Because as far as I can tell, `collectUsedFields` collects all fields used in a `@key`, that's even the first thing it does, but that's also what `keysPredicate` does, isn't it? Is there a subtle difference here I'm not seeing?

Reply:
The existing `collectUsedFields` will only collect fields that are marked with `@external`, so we also need to make sure that we're finding anything that is referenced by a key.

Reply:
Oh, I missed that part. Well, I think it's an issue, because for this feature we actually need to check the non-external field that are used, and this for `@provides`, `@requires` and interfaces (it's also confusing that it's called `isUsedField` and only return "external used field" versus "all used field").

What we should do is change `isUsedField` collect all used fields, and change the other consumers of `collectUsedFields` to filter non-external field out if they don't care about them. But for override, we explicitly care about only about the non-external used field.

Reply:
Ok, I understand now. I had stripped out the external restriction from the top level function, but not the helper function, so the caller is already doing the right thing. I'll change the naming and then get rid of the keys specific version.